pre_commit_hooks/string_fixer.py

```diff
@@ -32,7 +33,14 @@ def get_line_offsets_by_line_no(src):
 
 
 def fix_strings(filename):
-    contents = io.open(filename).read()
+    try:
+        contents = io.open(filename).read()
+    except Exception:
```

Comment:
Here's that in patch form:

```diff
diff --git a/pre_commit_hooks/string_fixer.py b/pre_commit_hooks/string_fixer.py
index 3523e8a..59b832f 100644
--- a/pre_commit_hooks/string_fixer.py
+++ b/pre_commit_hooks/string_fixer.py
@@ -11,15 +11,15 @@ double_quote_starts = tuple(s for s in tokenize.single_quoted if '"' in s)
 
 
 def handle_match(token_text):
-    if '"""' in token_text or "'''" in token_text:
+    if b'"""' in token_text or b"'''" in token_text:
         return token_text
 
     for double_quote_start in double_quote_starts:
         if token_text.startswith(double_quote_start):
             meat = token_text[len(double_quote_start):-1]
-            if '"' in meat or "'" in meat:
+            if b'"' in meat or b"'" in meat:
                 break
-            return double_quote_start.replace('"', "'") + meat + "'"
+            return double_quote_start.replace(b'"', b"'") + meat + b"'"
     return token_text
 
 
@@ -32,15 +32,15 @@ def get_line_offsets_by_line_no(src):
 
 
 def fix_strings(filename):
-    contents = io.open(filename).read()
+    contents = open(filename, 'rb').read()
     line_offsets = get_line_offsets_by_line_no(contents)
 
     # Basically a mutable string
-    splitcontents = list(contents)
+    splitcontents = bytearray(contents)
 
     # Iterate in reverse so the offsets are always correct
     tokens = reversed(list(tokenize.generate_tokens(
-        io.StringIO(contents).readline,
+        io.BytesIO(contents).readline,
     )))
     for token_type, token_text, (srow, scol), (erow, ecol), _ in tokens:
         if token_type == tokenize.STRING:
@@ -50,9 +50,9 @@ def fix_strings(filename):
                 line_offsets[erow] + ecol
             ] = new_text
 
-    new_contents = ''.join(splitcontents)
+    new_contents = bytes(splitcontents)
     if contents != new_contents:
-        with io.open(filename, 'w') as write_handle:
+        with open(filename, 'wb') as write_handle:
             write_handle.write(new_contents)
         return 1
     else:
diff --git a/tests/string_fixer_test.py b/tests/string_fixer_test.py
index a65213b..0603adf 100644
--- a/tests/string_fixer_test.py
+++ b/tests/string_fixer_test.py
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
 from __future__ import absolute_import
 from __future__ import print_function
 from __future__ import unicode_literals
@@ -49,3 +50,11 @@ def test_rewrite(input_s, output, expected_retval, tmpdir):
     retval = main([path.strpath])
     assert path.read() == output
     assert retval == expected_retval
+
+
+def test_non_utf8_file(tmpdir):
+    f_py = tmpdir.join('f.py')
+    f_py.write_binary('# -*- coding: cp1252 -*-\nx = "€"\n'.encode('cp1252'))
+    assert main((f_py.strpath,)) == 1
+    after = f_py.read_binary()
+    assert after == "# -*- coding: cp1252 -*-\nx = '€'\n".encode('cp1252')
```

This passes the tests in python 2, but I don't think `tokenize` works for bytes in python 3... hmm...  Instead the code could do the encoding dance (parse the comment cookie, assume utf-8 otherwise, decode to text, encode to the detected encoding) maybe?